mesonbuild/mesonlib/universal.py

```diff
@@ -2232,3 +2233,17 @@ def is_project(self) -> bool:
     def is_base(self) -> bool:
         """Convenience method to check if this is a base option."""
         return self.type is OptionType.BASE
+
+
+def first(iter: T.Iterable[_T], predicate: T.Callable[[_T], bool]) -> T.Optional[_T]:
+    """Find the first entry in an iterable where the given predicate is true
+
+    :param iter: The iterable to search
+    :param predicate: A finding function that takes an element from the iterable
+        and returns True if found, otherwise False
+    :return: The first found element, or None if it is not found
+    """
+    for i in iter:
+        if predicate(i):
+            return i
+    return None
```

Comment:
`next(i for i in iter if predicate(i))` or `next(filter(predicate, iter))` ?

Reply:
If there is no element though you get a `StopIteration` exception and have to catch it, so you actually end up with
```python
try:
    return next(i for i in iter if predicate(i))
except StopIteration:
    return None
````
Which doesn't seem any cleaner to me

Reply:
That's why `next()` has a `default` argument: `return next(i for i in iter if predicate(i), None)` 😃

Reply:
Is the goal here to do things in the least number of lines, or to be readable?

Reply:
My brain takes less time to parse the one liner than the for loop, but I guess it is subjective.

Reply:
The loop or the one-liner is the entirety of the function body, and it's a function to give the whole thing an interface that peoples' brains can parse in the least amount of time ("`first(foo, func)` gets the first thing in foo where func is true").

Accepting the usefulness of such a helper function, its main purpose is that people don't look at the loop at all, merely rely on its interface. That being the case, IMHO we should rely on whichever version of the function body guarantees understanding... and for me at least, one-liners are easier to read, but harder to be confident that I've read *correctly*.